src/validation.cpp

```diff
@@ -1717,6 +1717,15 @@ static bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockInd
                          REJECT_INVALID, "bad-blk-amount");
     }
 
+    // Masternode/Budget payments
+    // After transition to DMN system, this is checked also during IBD
+    if (!fInitialBlockDownload || deterministicMNManager->LegacyMNObsolete(pindex->nHeight)) {
+        if (!IsBlockPayeeValid(block, pindex->pprev)) {
+            mapRejectedBlocks.emplace(block.GetHash(), GetTime());
+            return state.DoS(0, false, REJECT_INVALID, "bad-cb-payee", false, "Couldn't find masternode/budget payment");
```

Comment:
This DoS can be a direct ban.
Atm, the only way to get here is if someone manually craft an invalid block paying to a different masternode/budget.

Reply:
This is still dependent on sporks, and it still needs mnsync for budget payments.
Have actually removed the check `deterministicMNManager->LegacyMNObsolete(pindex->nHeight)`, we must refactor it later, as mn payments are deterministic, but budget payments are not.

Reply:
kk